### PR TITLE
Add upgrade type verification function

### PR DESCRIPTION
**Pull request by chatton**
## Description

closes: #3446

### Commit Message / Changelog Entry

N/A feature branch

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [x] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

**Review by damiannolan — APPROVED** [review]

File: modules/core/04-channel/types/upgrade.go [hunk 1]
```diff
@@ -4,8 +4,35 @@ import (
 	errorsmod "cosmossdk.io/errors"
 
 	"github.com/cosmos/ibc-go/v7/internal/collections"
+	clienttypes "github.com/cosmos/ibc-go/v7/modules/core/02-client/types"
 )
 
+// NewUpgrade creates a new Upgrade instance.
+func NewUpgrade(modifiableFields ModifiableUpgradeFields, timeout UpgradeTimeout, lastPacketSent uint64) *Upgrade {
```
Comment (crodriguezvega) [inline modules/core/04-channel/types/upgrade.go:11]:
```suggestion
func NewUpgrade(upgradeFields UpgradeFields, timeout UpgradeTimeout, lastPacketSent uint64) *Upgrade {
```

**Review by crodriguezvega — APPROVED** [review]
Thank you, @chatton. Can we then remove `VerifyChannelUpgradeTimeout` and VerifyChannelUpgradeSequence` functions?

**Comment by chatton** [comment]
@crodriguezvega there should be a follow up issue that removes all the no longer required verification functions.

File: modules/core/03-connection/keeper/verify_test.go [hunk 2]
```diff
@@ -1020,6 +1020,85 @@ func (suite *KeeperTestSuite) TestVerifyUpgradeErrorReceiptAbsence() {
...
+			name: "fails when client state is frozen",
+			malleate: func() {
+				clientState := path.EndpointB.GetClientState().(*ibctm.ClientState)
+				clientState.FrozenHeight = clienttypes.NewHeight(0, 1)
+				path.EndpointB.SetClientState(clientState)
+			},
+			expPass: false,
+		},
+		{
+			name: "fails with bad client id",
```
[the remark this answers is not in the document]
Comment (chatton) [reply modules/core/03-connection/keeper/verify_test.go:1049]: we bave "bad client id" for all other tests doing this check, if we want to change it we should probably do it everywhere (I think not in this PR though!)

File: modules/core/03-connection/keeper/verify_test.go [hunk 3]
```diff
@@ -1020,6 +1020,85 @@ func (suite *KeeperTestSuite) TestVerifyUpgradeErrorReceiptAbsence() {
...
+		suite.Run(tc.name, func() {
+			suite.SetupTest() // reset
+
+			path = ibctesting.NewPath(suite.chainA, suite.chainB)
+			suite.coordinator.Setup(path)
+
+			upgrade = channeltypes.NewUpgrade(
+				channeltypes.NewModifiableUpgradeFields(channeltypes.UNORDERED, []string{path.EndpointA.ConnectionID}, "v1.0.0"),
+				channeltypes.NewUpgradeTimeout(clienttypes.ZeroHeight(), 100000),
+				0,
```
[the remark this answers is not in the document]
Comment (chatton) [reply modules/core/03-connection/keeper/verify_test.go:1078]: should last packet sent be 1 if we have not sent any packets?

File: modules/core/03-connection/keeper/verify.go [hunk 4]
```diff
@@ -364,16 +364,15 @@ func (k Keeper) VerifyNextSequenceRecv(
 	return nil
```
Comment (chatton) [inline modules/core/03-connection/keeper/verify.go:364]: the diff in this file is quite confusing, the following functions were removed

- `VerifyChannelUpgradeSequence`
- `VerifyChannelUpgradeTimeout`

**Review by damiannolan — APPROVED** [review]
ACK

**Pull request merged**